### PR TITLE
v8 5.8.48

### PR DESCRIPTION
**Pull request by oaleynik**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This defaults v8 to 5.8.48, targets `d8` only and allows build latest HEAD with --HEAD option

Based on: https://github.com/pinepain/homebrew-devtools/blob/master/Formula/v8%405.7.rb

### CODE REVIEW

**Comment by nijikon** [comment]
@BrewTestBot test this please

File: Formula/v8.rb
Comment (mathiasbynens) [inline Formula/v8.rb]: Could even do `bin.install "d8" => "v8"` IMHO. The `v8` shell in the code base is just an example with little real-world use — `d8` should be used instead. But it makes sense to give it a more appropriate name (just like `"ch" => "chakra"` when installing `brew install chakra`).

Comment (oaleynik) [inline Formula/v8.rb]: @mathiasbynens can I amend this commit and force push to the branch? Or it is ok to have two commits?

Comment (oaleynik) [inline Formula/v8.rb]: Indeed. I've been going to make that (d8 => v8), but was not sure is it ok :)

Comment (mathiasbynens) [inline Formula/v8.rb]: Since GitHub supports squash-and-merge nowadays, I’d assume Homebrew core devs are fine with either. Not sure though

Comment (oaleynik) [inline Formula/v8.rb]: @mathiasbynens pushed as separated commit. I can change it if will be needed :) Thanks for suggestion!

**Comment by pinepain** [comment]
5.8 breaks BC with 5.1 which is default now. Maybe let's use versioned formula -`v8@5.8`?

**Comment by pinepain** [comment]
**To clarify:** v8 releases doesn't follow semantic versioning at all: it even breaks BC in path releases, which happens on everyday basis, with [php-v8](https://github.com/pinepain/php-v8) I track v8 changes to public API almost on everyday basis. If there are packages who depends on default `v8`, they may stop working. Also, prior to 5.6 (or 5.7?) you had to build your binaries with v8 static  lib, which is not necessary nowadays as v8 build complete dynamic library.

I'd suggest just bring my [`v8@5.7`](https://github.com/pinepain/homebrew-devtools/blob/master/Formula/v8%405.7.rb) in, update it patch version to lkgr (last known good release) and creage `v8@5.8` from it with lkgr patch version. Later default `v8` could be phased out when in will be out of use (it could be tracked with homebrew build-in analytic and bintray downloads statistic).

P.S.: Apart of that - 👍, thanks for bringing it up!

**Comment by mathiasbynens** [comment]
Versioning is being discussed here: https://github.com/Homebrew/brew/issues/620

**Comment by pinepain** [comment]
@mathiasbynens I'm aware of, thanks. Do you think it would be better to move further discussion about this PR BC & versioning there?

**Comment by mathiasbynens** [comment]
@pinepain Thanks for your work on creating and maintaining V8 formulae!

[My comment](https://github.com/Homebrew/homebrew-core/pull/9343#issuecomment-275856360) wasn’t aimed at you specifically — it was meant as a useful remark for anyone reading this thread now that you’ve brought up versioning.

> Do you think it would be better to move further discussion about this PR BC & versioning there?

@ilovezfs already [said as much](https://github.com/Homebrew/homebrew-core/pull/7168#issuecomment-265132022), so yeah :) FWIW your proposal sounds 👍 to me.

**Comment by pinepain** [comment]
@mathiasbynens got it, good remark, last time @ilovezfs opened my eyes with poiting to that discussion which i somehow missed. Will bring it up there and create few PRs if we come up to some decision. 

As to exposing d8 as v8, I'd rather symlink or create some proxy script.

**Comment by pinepain** [comment]
> This defaults v8 to 5.8.48, targets d8 only

@oaleynik Does this formula builds all necessary headers and libs so that other devs can reuse it, or it **d8 only**, as you originally wrote?

If it d8-only, it limit it usage significantly and definitely breaks BC, at this time at least `php{5x,7x}-v8js` formulae depends on it.

**Comment by oaleynik** [comment]
@pinepain it is `d8` only. Intent is to have easy way for developers to check newest features in different JavaScript engines. It can make sense to create `v8-shell` formula which exposes `d8` as `v8` only.

Ref: https://github.com/mathiasbynens/todo/issues/17

**Comment by oaleynik** [comment]
IMHO for me it feels that `v8` formula should reference latest stable `V8`. At least I would expect to get latest `V8` if I run `brew install v8`

**Comment by pinepain** [comment]
@oaleynik 👍, one day it would be this way, but nowadays it may or even will unnecessary break things we may not even aware of. I'd suggest to go with some `v8-shell@x.y` which would depend on `v8@x.y` and expose `d8` binary as `v8` (see my note about ICU data file) or with version-independent  `v8-version` which would expose/switch current `d8/v8`.

FTR, latest stable is 5.6, not 5.8.

**Comment by oaleynik** [comment]
The idea is to have the way to check *latest* features. Not use the latest stable version. But I agree that `v8-shell` would be better as it clearly indicates what it exposes.

**Comment by pinepain** [comment]
@oaleynik I would also to have latest feature =), but v8 releases pop up almost every hour, sometimes even more often. You may want to build them with `--head`, or ask google v8 folks to package every it build for homebrew (if you will, ask please to build for ubuntu/debs and for centos/rhel families, internet won't forget that 😃). I'd prefer not to add --head option to version at this time, mostly because it's a bit cumbersome to get latest patch release version, but I will try to get it with `x.y-lkgr`, which should be far more enough for vast majority cases, normally lkgr is just a few patch versions behind (read, few hours behind). 

So will you be OK if I just PR my formula to homebrew-core (sure, adjusting to 5.8)? Then you can go with simple `v8-shell` (though, I'd suggest to make it versioned, `v8-shell@x.y` to prevent any misunderstandings or go with some d8 version switcher when multiple v8 versions installed).

**Comment by oaleynik** [comment]
@pinepain I agree with either. IMHO, before installing anything I would know what I need (dev deps or shell or whatever...) so I don't see an issue with having `up-to-date` version of V8 under `v8` formula. On the other hand, I think it will be correct to have formula named after its intent - `v8-devel`, `v8-shell`, etc.. too.

**Comment by pinepain** [comment]
@oaleynik have you looked what formulae depends on `v8`? Note, that homebrew-core is not the only tap, it is just a default one. Without being sure that nothing will be broken, I would not recommend to upgrade this core formula. Also notice, that because of internal changes to v8, js code run withing different v8 versions may provide different result, so if any depend on `v8` formula, they may be very surprised, but I would count just direct dependencies as the latter one is is incidental dependency and undetectable from out position.

P.S. having `--head` for 5.8 is really cumbersome as there is not lkgr yet, maybe it will be referenced later.

**Comment by oaleynik** [comment]
@pinepain in `homebrew-core` is see just one package which depends on `v8` https://github.com/Homebrew/homebrew-core/blob/0e6b8fcf56ca7976b1ae8489323771fed4e6813a/Formula/gjstest.rb

and one which has `v8` as optional:
https://github.com/Homebrew/homebrew-core/blob/496e1baa572036290610551bb7f5db19a87e27f0/Formula/uwsgi.rb

I'm not sure about all the taps around

**Comment by MikeMcQuaid** [comment]
> oaleynik commented 22 hours ago • edited
> @pinepain in homebrew-core I see just one package which depends on v8 https://github.com/Homebrew/homebrew-core/blob/0e6b8fcf56ca7976b1ae8489323771fed4e6813a/Formula/gjstest.rb
> 
> and one which has v8 as optional:
> https://github.com/Homebrew/homebrew-core/blob/496e1baa572036290610551bb7f5db19a87e27f0/Formula/uwsgi.rb
> 
> I'm not sure about all the taps around

As long as we don't break these I'm fine with updating the V8 version here.

File: Formula/v8.rb
Comment (MikeMcQuaid) [inline Formula/v8.rb]: Leave this as-is, it'll be updated when we merge.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by MikeMcQuaid** [comment]
Various CI failures: 
- https://bot.brew.sh/job/Homebrew%20Core/15494/version=el_capitan/testReport/junit/brew-test-bot/el_capitan/audit_v8___online/
- https://bot.brew.sh/job/Homebrew%20Core/15494/version=el_capitan/testReport/junit/brew-test-bot/el_capitan/style_v8/
- https://bot.brew.sh/job/Homebrew%20Core/15494/version=el_capitan/testReport/junit/brew-test-bot/el_capitan/test_gjstest/

**Comment by pinepain** [comment]
@MikeMcQuaid this PR is fundamentally broken, not even mention that it breaks dependencies (not only listed one).

I would vote to use my PR #9392 which also provides working d8  (d8 from this PR won't work properly because of how ICU data file loaded by default, in referenced PR I patched that).

**Comment by MikeMcQuaid** [comment]
> @MikeMcQuaid this PR is fundamentally broken, not even mention that it breaks dependencies (not only listed one).

@pinepain It may be but it can be improved. I'll comment on your PR.

**Comment by pinepain** [comment]
@MikeMcQuaid sure it could be, but even if we copy the formula from my PR we can't do such update in `v8` formula itself as it will break all deps. V8 changed significantly between 5.1 and 5.8 in many aspects, from breaking API/ABI to building process.

**Comment by MikeMcQuaid** [comment]
> as it will break all deps

We should try to fix these deps and report issues to upstream projects. In this PRs case it only seems to break a single formula.

**Comment by pinepain** [comment]
@MikeMcQuaid this PR drop all .dylib building, so any deps won't work with it. As well as relocating d8 will lead to semi-working shell (see my comments above). Also, as you mentioned, 5.8 is not a stable (5.6 is). But @oaleynik interested in having latest v8 **shell** with latest ES features, if I get it right. While I'm absolutely for this changes, I'd rather keep them in a separate version, like I submitted.

> We should try to fix these deps and report issues to upstream projects
it might take from weeks to months, as a lot of changed in V8 API between 5.1 to 5.8 (I work with V8 actively since 4.8).

**Comment by MikeMcQuaid** [comment]
> it might take from weeks to months, as a lot of changed in V8 API between 5.1 to 5.8 (I work with V8 actively since 4.8).

That's fine. That's our typical process for major upgrades like this.

**Comment by MikeMcQuaid** [comment]
> this PR drop all .dylib building, so any deps won't work with it.

We should continue to build dylibs or have dependencies build statically.

**Comment by pinepain** [comment]
OK,  let's sum up what we want to achieve:
 - do we want to upgrade v8 to 5.8 in this formula? 
 - do we want to upgrade v8 to just a stable version?

**Comment by MikeMcQuaid** [comment]
If we upgrade v8 it should be to a stable version. We could consider adding 5.8 as a `devel` block regardless.

**Comment by pinepain** [comment]
@oaleynik ftr, I added [v8@5.8](https://github.com/pinepain/homebrew-devtools/blob/master/Formula/v8%405.8.rb) to [pinepain/devtools](https://github.com/pinepain/homebrew-devtools), so if you up to give it a try, just `brew tap pinepain/devtools` && `brew install v8@5.8`. The `d8` exposed as `v8`, the only thing you will need to do, is to add opt_bin path to your `$PATH`.  I tested it and I guess it's what you need. Any later improvements are highly welcomed and thank you a lot for bringing the issue up! You did enormous work! I know how it time consuming to play around v8 and how many hidden snags does it have, so thank you one more time.

**Comment by MikeMcQuaid** [comment]
> What is the main issue having v8@5.8 versioned formula?

It'll have to be maintained indefinitely and it's not a stable release which is against our packaging guidelines.

**Comment by oaleynik** [comment]
@MikeMcQuaid I've cleared all the audit errors. This is my first PR to Homebrew - I've missed initially that I can't run audit on my fork (or probably I've done it wrong). Branch has been rebased over latest upstream master and all intermediary commits have been squashed. Also, I've set version to 5.6.326 which is the version of V8 used in Chrome stable. It is still possible to build from HEAD. @pinepain there is no `5.8-lgkr` tag or branch for V8 FYI.

@pinepain Although I found your https://github.com/Homebrew/homebrew-core/pull/9343#issuecomment-275927404 a bit insulting and disrespectful, which can turndown any desire to continue contributing to Homebrew - I highly appriciate all your help and suggestions. I believe that you have not had in mind any means to attack somebody or be aggressive. As I said a few comments above your concerns regarding such an optimistic update of v8 are reasonable. But I also think that having latest (well, at least latest stable with an ability to build from HEAD) V8 under `v8` formula will be correct as well. 

I'm not at the position of making the final decision here, however I believe that all together we can find the optimal solution!

**Comment by pinepain** [comment]
@oaleynik I'm sorry that you found it in this way, though it was targets not a author personality, but the profound effect of what this changes will make on all dependencies.

It you want to contribute, just take formula from my pr, rename formula class to `V8` and replace version with 5.6.326.42. If I get you right, you need working d8, if it is, that's probably what you want. Thought, it would be just a 5.6.

**Comment by pinepain** [comment]
> It'll have to be maintained indefinitely and it's not a stable release which is against our packaging guidelines.

@MikeMcQuaid yeah, makes sense, then I'll keep it in a separate tap as before.

**Comment by MikeMcQuaid** [comment]
> @oaleynik I'm sorry that you found it in this way, though it was targets not your personality, but the profound effect of what this changes will make on all dependencies.

Just FYI: "I'm sorry that you X" is not an apology (it doesn't imply your actions are at fault but the other person was) and I agree that there's definitely more diplomatic ways of phrasing things than "fundamentally broken". Let's all try and assume the best in each other: everyone here is trying to make Homebrew better.

> @MikeMcQuaid yeah, makes sense, then I'll keep it in a separate tap as before.

If you're not going to be making your own bottles then a `devel` build here is probably just as good but ultimately: up to you.

**Comment by pinepain** [comment]
To prevent further misunderstanding, I'll drop few lines in Russian.

@oaleynik Олег, я действительно не хотел ничего такого обидного написать, извини. Основноая идея в первоначальном виде данные измменения реально ломает очень много чего, т.е. не просто зависимости он ломает, а даже концепцию данной формулы. d8 просто консольная утилита, к слову, довольно примитивная если сравнивать с нодой ([src](https://github.com/v8/v8/blob/master/src/d8.cc), в то время как осоновная логика здесь в динамических библиотеках, сборка которых и выхлоп собственно и очень сильно помемнялась. Раньше собирались статические и динамические библиотеки, но примерно в 5.4 в v8 наконец-то довернули систему сборки и статические стали ненужны. И если для зависимостей решить проблему сборки путем исключения линкования статических библиотек не есть большая проблема, то вот тот ворох изменений в API v8 вряд ли кто-то будет разгребать. Вот собствено что я и имел в виду, но изложил из рук вон плохо. Впредь я таки лучше распишу личний раз чем просто писать "fundamentally broken".

@MikeMcQuaid:
 - Can we depends on --devel from other formula?
 - Is it ok to bump v8 version in devel?
 - Are there any chance that homebrew will bottle devel version in forseable future? Are there any discussions or some other resources where I can read.

**Comment by MikeMcQuaid** [comment]
> Can we depends on --devel from other formula?

We cannot. That's another good case for a separate formula in your tap (until we need it in homebrew-core).

> Is it ok to bump v8 version in devel?

Yep.

> Are there any chance that homebrew will bottle devel version in forseable future? Are there any discussions or some other resources where I can read.

If someone submits a PR for it: yep. There's no active plans for feature issue tracking this currently.

**Pull request closed without merging**
